File: SSUMobile/Modules/Core/Module/SSUModuleServices.m
//
//  SSUModuleServices.m
//  SSUMobile
//
//  Created by Eric Amorde on 08/17/2016.
//  Copyright © 2016 Sonoma State University Department of Computer Science. All rights reserved.
//

#import "SSUModuleServices.h"
#import "SSUConfiguration.h"

NSString * const SSUModulesEnabledKey = @"edu.sonoma.modules.enabled";
NSString * const SSUModulesDidLoadNotification = @"edu.sonoma.modules.loaded.notification";

@interface SSUModuleServices()

/** All SSUModule(UI) instances that were loaded on launch */
@property (nonatomic, readwrite) NSArray * modules;
/** All SSUModuleUI instances */
@property (nonatomic, readwrite) NSArray * modulesUI;

@end

@implementation SSUModuleServices

+ (instancetype) sharedInstance {
    static SSUModuleServices * instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[self alloc] init];
    });
    return instance;
}

- (void) loadModules {
    NSMutableArray * moduleObjects = (id)[NSMutableArray new];
    for (NSString * className in [self moduleClasses]) {
        Class cls = NSClassFromString(className);
        SSUModuleBase<SSUModule> * module = [cls sharedInstance];
        NSAssert([module conformsToProtocol:@protocol(SSUModule)], @"Module with name %@ does not conform to the %@ protocol", className, NSStringFromProtocol(@protocol(SSUModule)));
        [moduleObjects addObject:module];
    }
    
    self.modules = moduleObjects;
    self.modulesUI = [self modulesConformingToProtocol:@protocol(SSUModuleUI)];
    [[NSNotificationCenter defaultCenter] postNotificationName:SSUModulesDidLoadNotification object:self.modules];
}

- (void) reloadModules {
    self.modules = nil;
    self.modulesUI = nil;
}

- (NSArray *) moduleClasses {
    NSArray <NSString *> * moduleClasses = [[SSUConfiguration sharedInstance] stringArrayForKey:@"edu.sonoma.modules.enabled"];
#ifdef DEBUG
    if (NSClassFromString(@"SSUDebugModule")) {
        return [moduleClasses arrayByAddingObject:@"SSUDebugModule"];
    }
    return moduleClasses;
#else
    return moduleClasses;
#endif
}

- (NSArray *) modules {
    if (_modules) return _modules;
    
    NSMutableArray * moduleObjects = (id)[NSMutableArray new];
    for (NSString * className in [self moduleClasses]) {
        Class cls = NSClassFromString(className);
        SSUModuleBase<SSUModule> * module = [cls sharedInstance];
        NSAssert([module conformsToProtocol:@protocol(SSUModule)], @"Module with name %@ does not conform to the %@ protocol", className, NSStringFromProtocol(@protocol(SSUModule)));
        [moduleObjects addObject:module];
    }
    
    _modules = [moduleObjects copy];
    
    return _modules;
}

- (NSArray *) modulesUI {
    if (_modulesUI) return _modulesUI;
    
    NSMutableArray * moduleObjects = [NSMutableArray new];
    for (SSUModuleBase<SSUModule> * module in self.modules) {
        if ([module conformsToProtocol:@protocol(SSUModuleUI)]) {
            [moduleObjects addObject:module];
        }
    }
    
    _modulesUI = (id)[self modulesConformingToProtocol:@protocol(SSUModuleUI)];
    
    return _modulesUI;
}

- (id<SSUModule>) moduleWithIdentifier:(NSString *)identifier {
    for (SSUModuleBase<SSUModule> * module in self.modules) {
        if ([module.identifier isEqualToString:identifier]) {
            return module;
        }
    }
    return nil;
}

- (NSArray *) modulesConformingToProtocol:(Protocol *)protocol {
    NSMutableArray * conformingModules = [NSMutableArray new];
    for (id<SSUModule> module in self.modules) {
        if ([module conformsToProtocol:protocol]) {
            [conformingModules addObject:module];
        }
    }
    return conformingModules;
}

- (void) updateAll {
    for (id<SSUModule> module in self.modules) {
        [module updateData:NULL];
    }
}

- (void) setupAll {
    for (id<SSUModule> module in self.modules) {
        [module setup];
    }
}


@end

File: SSUMobile/Modules/Core/Module/SSUModuleServices.swift
//
//  SSUModuleServices.swift
//  SSUMobile
//
//  Created by Eric Amorde on 4/7/17.
//  Copyright © 2017 Sonoma State University Department of Computer Science. All rights reserved.
//

import Foundation

let SSUModulesEnabledKey = "edu.sonoma.modules.enabled"
let SSUModulesDidLoadNotification = Notification.Name("edu.sonoma.modules.loaded.notification")

class SSUModuleServices: NSObject {
    
    @objc(sharedInstance)
    static let instance = SSUModuleServices()
    
    override init() {}
    
    var modules: [SSUModule] = []
    var modulesUI: [SSUModuleUI] {
        return modules.flatMap({ $0 as? SSUModuleUI })
    }
    
    /**
     List of all supported modules which may or may not be included based on feature flags
     */
    private var supportedModules: [SSUModule] = {
        var result: [SSUModule] = [
            SSUAboutModule.instance,
            SSUCalendarModule.instance,
            SSUDirectoryModule.instance,
            SSUEmailModule.instance,
            SSUNewsModule.instance,
            SSUMapModule.instance,
            SSUResourcesModule.instance,
            SSURadioModule.instance
        ]
        #if DEBUG
        result.append(SSUDebugModule.instance)
        #endif
        return result
    }()
    
    private var identifiers: [String] {
        var ids = SSUConfiguration.sharedInstance().stringArray(forKey: SSUModulesEnabledKey) ?? []
        #if DEBUG
        ids.append("debug")
        #endif
        return ids
    }
    
    func addModule(_ module: SSUModule) {
        modules.append(module)
    }
    
    func loadModules() {
        modules = []
        for name in identifiers {
            if let module = supportedModules.first(where: { $0.identifier == name }) {
                addModule(module)
            }
        }
        
        NotificationCenter.default.post(name: SSUModulesDidLoadNotification, object: modules)
    }
    
    func modulesConformingTo<T>(type: T.Type) -> [T] {
        return modules.flatMap({ $0 as? T })
    }
    
    @objc
    func modulesConformingToProtocol(_ proto: Protocol) -> [Any] {
        return modules.filter({ $0.conforms(to: proto) })
    }
    
    func getModule(identifier: String) -> SSUModule? {
        return modules.first(where: { $0.identifier == identifier })
    }
    
    func updateAll() {
        modules.forEach({ $0.updateData?(nil) })
    }
    
    func setupAll() {
        modules.forEach({ $0.setup() })
    }
}
